src/transformers/generation/stopping_criteria.py

```diff
@@ -126,6 +129,167 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
         return time.time() - self.initial_timestamp > self.max_time
 
 
+class StopStringCriteria(StoppingCriteria):
+    """
+    This class can be used to stop generation whenever specific string sequences are encountered. It preprocesses
+    the strings together with the tokenizer vocab to find positions where tokens can validly complete the stop strings.
+
+    Args:
+        tokenizer (`PreTrainedTokenizer`):
+            The model's associated tokenizer (necessary to extract vocab and tokenize the termination sequences)
+        stop_strings (`Union[str, List[str]]`):
+            A list of strings that should end generation. If a string is passed, it will be treated like a
+            list with a single element.
+    """
+
+    def __init__(self, tokenizer: PreTrainedTokenizerBase, stop_strings: Union[str, List[str]]):
+        if isinstance(stop_strings, str):
+            stop_strings = [stop_strings]
+
+        self.vocab = tokenizer.get_vocab()
+        self.stop_strings: List[str] = stop_strings
+        self.strings_to_valid_positions, self.strings_to_end_lengths = self.get_matching_positions(
+            self.vocab, stop_strings
+        )
+
+        self.max_valid_positions = {
+            stop_string: max([len(val) for val in self.strings_to_valid_positions[stop_string].values()])
+            for stop_string in stop_strings
+        }
+        self.max_valid_end_lens = {
+            stop_string: max([len(val) for val in self.strings_to_end_lengths[stop_string].values()])
+            for stop_string in stop_strings
+        }
+        self.embedding_vecs = self.create_embedding_vecs()
+
+    @staticmethod
+    def get_matching_positions(vocab, stop_strings):
+        """This function preprocesses stop strings and the tokenizer vocabulary to determine where tokens can
+        validly appear in the stop strings. For each stop string, it returns a dictionary mapping tokens to a list of
+        valid positions, as well as a dictionary mapping tokens to a list of possible overlap lengths at the
+        end of the stop string."""
+        tok_list = list(vocab.keys())
+        strings_to_valid_positions = {}
+        strings_to_end_lengths = {}
+        for stop_string in stop_strings:
+            strings_to_valid_positions[stop_string] = {}
+            strings_to_end_lengths[stop_string] = {}
+            for token in tok_list:
+                matching_positions = []
+                possible_end_lengths = []
+                for i in range(1 - len(token), len(stop_string)):
+                    tok = token[::-1].replace("▁", " ")
```

Comment:
Ah, this is a really good point! I forgot that different tokenizers have different ways of encoding spaces. I should look for a correct general method to convert tokens to strings instead of writing all of the different special cases